Mock unsafe traits

I have been wanting to program something ui, with rust.
I have been using vulkano and winit.

I now want to add a unit-test.  The Window creation boils down, to implementing this trait:
[link](https://github.com/rust-windowing/raw-window-handle/blob/cb628db982201c4c6a97ea56fb117e739cbc9c38/src/lib.rs#L52)
```rust
pub unsafe trait HasRawWindowHandle {
    fn raw_window_handle(&self) -> RawWindowHandle;
}
```

I am trying, to mock this, by doing the following:
```rust
        mock! {
            Window{}
            unsafe impl raw_window_handle::HasRawWindowHandle for Window{
                fn raw_window_handle(&self) -> raw_window_handle::RawWindowHandle;
            }
        }
```
The compiler just says the following: ```error[E0200]: the trait `HasRawWindowHandle` requires an `unsafe impl` declaration```
I thought, that I specified unsafe impl ? Is this supported?

EDIT: The unsafe in the mock! macro is ignored when the macro expands.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,13 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+## [ Unreleased ] - ReleaseDate
+
+### Added
+
+- `mock!` and `#[automock]` now support `unsafe` traits.
+  ([#313](https://github.com/asomers/mockall/pull/313))
+
 ## [ 0.10.1 ] - 2021-07-01
 
 ## Fixed

--- a/mockall/tests/automock_unsafe_trait.rs
+++ b/mockall/tests/automock_unsafe_trait.rs
@@ -1,0 +1,36 @@
+// vim: ts=80
+#![deny(warnings)]
+
+use mockall::*;
+
+#[automock]
+pub unsafe trait Foo {
+    fn foo(&self) -> i32;
+}
+
+struct Baz{}
+
+#[automock]
+unsafe impl Foo for Baz {
+    fn foo(&self) -> i32 {
+        unimplemented!()
+    }
+}
+
+#[test]
+fn automock_trait() {
+    let mut mock = MockFoo::new();
+    mock.expect_foo()
+        .return_const(42);
+
+    assert_eq!(42, mock.foo());
+}
+
+#[test]
+fn automock_trait_impl() {
+    let mut mock = MockBaz::new();
+    mock.expect_foo()
+        .return_const(42);
+
+    assert_eq!(42, mock.foo());
+}

--- a/mockall/tests/mock_unsafe_trait.rs
+++ b/mockall/tests/mock_unsafe_trait.rs
@@ -1,0 +1,24 @@
+// vim: ts=80
+#![deny(warnings)]
+
+use mockall::*;
+
+pub unsafe trait Bar {
+    fn bar(&self) -> i32;
+}
+
+mock! {
+    pub Foo{}
+    unsafe impl Bar for Foo {
+        fn bar(&self) -> i32;
+    }
+}
+
+#[test]
+fn return_const() {
+    let mut mock = MockFoo::new();
+    mock.expect_bar()
+        .return_const(42);
+
+    assert_eq!(42, mock.bar());
+}

--- a/mockall_derive/src/mock_trait.rs
+++ b/mockall_derive/src/mock_trait.rs
@@ -28,7 +28,8 @@ pub(crate) struct MockTrait {
     /// Path on which the trait is implemented.  Usually will be the same as
     /// structname, but might include concrete generic parameters.
     self_path: PathSegment,
-    pub types: Vec<ImplItemType>
+    pub types: Vec<ImplItemType>,
+    pub unsafety: Option<Token![unsafe]>
 }
 
 impl MockTrait {
@@ -116,7 +117,8 @@ impl MockTrait {
             ss_name,
             trait_path,
             self_path,
-            types
+            types,
+            unsafety: impl_.unsafety
         }
     }
 
@@ -156,9 +158,10 @@ impl MockTrait {
         let trait_path = &self.trait_path;
         let self_path = &self.self_path;
         let types = &self.types;
+        let unsafety = &self.unsafety;
         quote!(
             #(#trait_impl_attrs)*
-            impl #ig #trait_path for #self_path #wc {
+            #unsafety impl #ig #trait_path for #self_path #wc {
                 #(#consts)*
                 #(#types)*
                 #(#calls)*

